protocols/bftcosi/bftcosi.go

```diff
@@ -187,291 +200,300 @@ func (bft *ProtocolBFTCoSi) Dispatch() error {
-// commitment round.
-func (bft *ProtocolBFTCoSi) startCommitmentCommit() error {
-	cm := bft.commit.CreateCommitment()
+// handleCommitment collects all commitments from children and passes them
+// to the parent or starts the challenge-round if it's the root.
+func (bft *ProtocolBFTCoSi) handleCommitment(comm Commitment) error {
+	dbg.Lvl4(bft.Name(), "RoundType:", comm.TYPE)
 
-	dbg.Lvl4(bft.Name(), "BFTCoSi Start Commitment COMMIT")
-	return bft.SendToParent(&Commitment{TYPE: RoundCommit, Commitment: cm})
-}
-
-// handle the arrival of a commitment
-func (bft *ProtocolBFTCoSi) handleCommit(comm Commitment) error {
-	typedCommitment := &Commitment{}
-	var commitment *cosi.Commitment
+	var commitment abstract.Point
 	// store it and check if we have enough commitments
+	bft.tpcMut.Lock()
+	defer bft.tpcMut.Unlock()
 	switch comm.TYPE {
 	case RoundPrepare:
-		bft.tpcMut.Lock()
 		bft.tempPrepareCommit = append(bft.tempPrepareCommit, comm.Commitment)
 		if len(bft.tempPrepareCommit) < len(bft.Children()) {
-			bft.tpcMut.Unlock()
 			return nil
 		}
-		commitment = bft.prepare.Commit(bft.tempPrepareCommit)
-		bft.tpcMut.Unlock()
+		commitment = bft.prepare.Commit(nil, bft.tempPrepareCommit)
 		if bft.IsRoot() {
-			return bft.startChallengePrepare()
+			return bft.startChallenge(RoundPrepare)
 		}
-
-		dbg.Lvl4(bft.Name(), "BFTCoSi handle Commit PREPARE")
 	case RoundCommit:
-		bft.tccMut.Lock()
 		bft.tempCommitCommit = append(bft.tempCommitCommit, comm.Commitment)
 		if len(bft.tempCommitCommit) < len(bft.Children()) {
-			bft.tccMut.Unlock()
 			return nil
 		}
-		commitment = bft.commit.Commit(bft.tempCommitCommit)
-		bft.tccMut.Unlock()
+		commitment = bft.commit.Commit(nil, bft.tempCommitCommit)
 		if bft.IsRoot() {
 			// do nothing:
 			// stop the processing of the round, wait the end of
 			// the "prepare" round: calls startChallengeCommit
 			return nil
 		}
 
-		dbg.Lvl4(bft.Name(), "BFTCoSi handle Commit COMMIT")
 	}
 	// set same RoundType as for the received commitment:
-	typedCommitment.TYPE = comm.TYPE
-	typedCommitment.Commitment = commitment
+	typedCommitment := &Commitment{
+		TYPE:       comm.TYPE,
+		Commitment: commitment,
+	}
 	return bft.SendToParent(typedCommitment)
 }
 
-// startPrepareChallenge create the challenge and send its down the tree
-func (bft *ProtocolBFTCoSi) startChallengePrepare() error {
-	// create challenge from message's hash:
-	hash := bft.Suite().Hash()
-	hash.Write(bft.Msg)
-	h := hash.Sum(nil)
-
-	ch, err := bft.prepare.CreateChallenge(h)
-	if err != nil {
-		return err
-	}
-
-	bftChal := &ChallengePrepare{
-		TYPE:      RoundPrepare,
-		Challenge: ch,
-		Msg:       bft.Msg,
+// startChallenge creates the challenge and sends it to its children
+func (bft *ProtocolBFTCoSi) startChallenge(t RoundType) error {
+	if t == RoundPrepare {
+		// create challenge from message's hash:
+		// XXX Why ? Since Challenge() already hash the message
```

Comment:
No discussion in the code please - it's difficult to follow. Please use github-issues or PRs. In that special case we should ask Lefteris.